metals/src/main/scala/scala/meta/internal/metals/BuildTargetClassesFinder.scala

```diff
@@ -0,0 +1,111 @@
+package scala.meta.internal.metals
+
+import ch.epfl.scala.{bsp4j => b}
+import scala.util.Failure
+import scala.util.Success
+import scala.util.Try
+
+class BuildTargetClassesFinder(
+    buildTargets: BuildTargets,
+    buildTargetClasses: BuildTargetClasses
+) {
+
+  //In case of success returns non-empty list
+  def findMainClassAndItsBuildTarget(
+      className: String,
+      buildTarget: Option[String]
+  ): Try[List[(b.ScalaMainClass, b.BuildTarget)]] =
+    buildTarget.fold {
+      val classes =
+        buildTargetClasses.findMainClassByName(className).collect {
+          case (clazz, BuildTargetIdOf(buildTarget)) => (clazz, buildTarget)
+        }
+      if (classes.nonEmpty) Success(classes)
+      else
+        Failure(new ClassNotFoundException(className))
+    } { targetName =>
+      buildTargets
+        .findByDisplayName(targetName)
+        .fold[Try[List[(b.ScalaMainClass, b.BuildTarget)]]] {
+          Failure(
+            new BuildTargetNotFoundException(targetName)
+          )
+        } { target =>
+          buildTargetClasses
+            .classesOf(target.getId())
+            .mainClasses
+            .values
+            .find(
+              _.getClassName == className
+            )
+            .fold[Try[List[(b.ScalaMainClass, b.BuildTarget)]]] {
+              Failure(
+                new ClassNotFoundInBuildTargetException(
+                  className,
+                  targetName
+                )
+              )
+            } { clazz => Success(List(clazz -> target)) }
+        }
+    }
+
+  //In case of success returns non-empty list
+  def findTestClassAndItsBuildTarget(
+      className: String,
+      buildTarget: Option[String]
+  ): Try[List[(String, b.BuildTarget)]] =
+    buildTarget.fold {
+      val classes =
+        buildTargetClasses.findTestClassByName(className).collect {
+          case (clazz, BuildTargetIdOf(buildTarget)) => (clazz, buildTarget)
+        }
+      if (classes.nonEmpty) Success(classes)
+      else
+        Failure(new ClassNotFoundException(className))
+    } { targetName =>
+      buildTargets
+        .findByDisplayName(targetName)
+        .fold[Try[List[(String, b.BuildTarget)]]] {
+          Failure(
+            new BuildTargetNotFoundException(targetName)
+          )
+        } { target =>
+          buildTargetClasses
+            .classesOf(target.getId())
+            .testClasses
```

Comment:
This looks like the only difference between the 2 functions is here and two lines below. Can w reuse the code and just add maybe a function `Classes => Option[T]`?

Reply:
okay, I've added
```
private def findClassAndBuildTarget[A](
      className: String,
      buildTarget: Option[String],
      findClassesByName: String => List[(A, b.BuildTargetIdentifier)],
      classesByBuildTarget: b.BuildTargetIdentifier => Iterable[A],
      getClassName: A => String
  ): Try[List[(A, b.BuildTarget)]]
```
to generalize them